### PR TITLE
make localstack possible

### LINKED ISSUE #206: Invalid security token when running against localstack
### What happened?

When running the provider against localstack, I'm seeing the following message:
```
cannot get terraform setup: cannot get the caller identity: GetCallerIdentity query failed: GetCallerIdentity query failed: operation error STS: GetCallerIdentity, https response error StatusCode: 403, RequestID: 821527d0-d6fa-441d-b5d3-a280b4767650, api error InvalidClientTokenId: The security token included in the request is invalid.
```

### How can we reproduce it?

1. Create a cluster with kind
2. Start localstack with `localstack start -d`
3. Connect localstack's container with `docker network connect kind localstack_main`
4. Install crossplane via Helm: `helm upgrade --install crossplane --create-namespace --namespace crossplane-system crossplane-stable/crossplane --wait > /dev/null 2>&1`
5. Create provider:
```
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: default
spec:
  package: xpkg.upbound.io/upbound/provider-aws:v0.21.0
```
6. Get localstack's container IP via `docker inspect localstack_main | jq -r ".[].NetworkSettings.Networks[\"kind\"].IPAddress"`
7. Provision `ProviderConfig` with the IP from previous step:
```
apiVersion: aws.upbound.io/v1beta1
kind: ProviderConfig
metadata:
  name: default
spec:
  endpoint:
    url:
      type: Static
      static: http://${LOCALSTACK_IP}:4566
  credentials:
    source: Secret
    secretRef:
      namespace: crossplane-system
      name: aws-creds
      key: creds
---
apiVersion: v1
kind: Secret
type: Opaque
metadata:
  name: aws-creds
  namespace: crossplane-system
stringData:
  creds: |
    [default]
    aws_access_key_id = test
    aws_secret_access_key = test
```
8. Provision some bucket:
```
apiVersion: s3.aws.upbound.io/v1beta1
kind: Bucket
metadata:
  labels:
    testing.upbound.io/example-name: upbound-example
  name: somebucket 
spec:
  forProvider:
    forceDestroy: true
    region: us-west-1
```

When trying this with [crossplane-contrib/provider-aws](https://github.com/crossplane-contrib/provider-aws), it works fine, which makes me think it might have to do with the provider itself. I can't see any request getting through to localstack either.

### What environment did it happen in?

* Universal Crossplane Version: N/A
* Provider Version: 0.21.0

### PR DESCRIPTION
**Pull request by alexinthesky**
### Description of your changes
updates providerconfig CRD and generated terraform provider block to meet localstack needs ( see https://docs.localstack.cloud/user-guide/integrations/terraform/#endpoint-configuration )

Fixes #206 

I have:

- [X] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
ran on local crossplane install with localstack.

### CODE REVIEW

**Comment by patelronak** [comment]
@ulucinar would love to get this reviewed and pushed out for my migration work from contrib aws provider to upbound.

**Comment by jeanduplessis** [comment]
@patelronak I just want to manage expectations here in that most Upbound employees are at our annual company offsite next week so a response may be delayed here.

**Comment by shaundewberry** [comment]
Implemented @alexinthesky's changes to provider-aws and built and deployed the provider locally in a kind cluster for successful creation of s3 bucket in localstack. Adding the new parameters in the ProviderConfig allowed bypass of the GetCallerIdentity error.

File: apis/v1beta1/types.go [hunk 1]
```diff
@@ -21,6 +21,17 @@ type ProviderConfigSpec struct {
 	// of AWS calls made by the provider.
 	// +optional
 	Endpoint *EndpointConfig `json:"endpoint,omitempty"`
+	// Whether to skip credentials validation via the STS API.
+	// This can be useful for testing and for AWS API implementations that do not have STS available.
+	SkipCredsValidation bool `json:"skip_credentials_validation,omitempty"`
```
Comment (sergenyalcin) [inline apis/v1beta1/types.go:26]: I think this is an optional field. Could you please the `+optional` mark for this?

File: apis/v1beta1/types.go [hunk 2]
```diff
@@ -21,6 +21,17 @@ type ProviderConfigSpec struct {
...
+	// Whether to enable the request to use path-style addressing, i.e., https://s3.amazonaws.com/BUCKET/KEY.
+	S3UsePathStyle bool `json:"s3_use_path_style,omitempty"`
```
Comment (sergenyalcin) [inline apis/v1beta1/types.go:28]: I think this is an optional field. Could you please the `+optional` mark for this?

File: apis/v1beta1/types.go [hunk 3]
```diff
@@ -21,6 +21,17 @@ type ProviderConfigSpec struct {
...
+	// Whether to skip the AWS Metadata API check
+	// Useful for AWS API implementations that do not have a metadata API endpoint.
+	SkipMetadataApiCheck bool `json:"skip_metadata_api_check,omitempty"`
```
Comment (sergenyalcin) [inline apis/v1beta1/types.go:31]: I think this is an optional field. Could you please the `+optional` mark for this?

File: apis/v1beta1/types.go [hunk 4]
```diff
@@ -21,6 +21,17 @@ type ProviderConfigSpec struct {
...
+	// Whether to skip requesting the account ID.
+	// Useful for AWS API implementations that do not have the IAM, STS API, or metadata API
+	SkipReqAccountId bool `json:"skip_requesting_account_id,omitempty"`
```
Comment (sergenyalcin) [inline apis/v1beta1/types.go:34]: I think this is an optional field. Could you please the `+optional` mark for this?

File: internal/clients/aws.go [hunk 5]
```diff
@@ -39,27 +52,44 @@ func TerraformSetupBuilder(version, providerSource, providerVersion string) terr
 		if err != nil {
 			return terraform.Setup{}, errors.Wrap(err, "failed to retrieve aws credentials from aws config")
 		}
-		identity, err := GlobalCallerIdentityCache.GetCallerIdentity(ctx, *cfg, creds)
-		if err != nil {
-			return terraform.Setup{}, errors.Wrap(err, "cannot get the caller identity")
+		AccountId := "000000000"
```
Comment (sergenyalcin) [inline internal/clients/aws.go:55]:
```suggestion
		accountId := "000000000"
```

File: internal/clients/aws.go [hunk 6]
```diff
@@ -39,27 +52,44 @@ func TerraformSetupBuilder(version, providerSource, providerVersion string) terr
...
+		if !pc.Spec.SkipCredsValidation {
+			identity, err := GlobalCallerIdentityCache.GetCallerIdentity(ctx, *cfg, creds)
+			if err != nil {
+				return terraform.Setup{}, errors.Wrap(err, "cannot get the caller identity")
+			}
+			AccountId = *identity.Account
```
Comment (sergenyalcin) [inline internal/clients/aws.go:61]:
```suggestion
			accountId = *identity.Account
```

File: internal/clients/aws.go [hunk 7]
```diff
@@ -39,27 +52,44 @@ func TerraformSetupBuilder(version, providerSource, providerVersion string) terr
...
+				keySecretAccessKey:      creds.SecretAccessKey,
+				keySessionToken:         creds.SessionToken,
+				keyEndpoints:            endpoints,
 			},
+
 			// Account ID is not part of provider configuration schema, so it
 			// needs to be given separately.
 			ClientMetadata: map[string]string{
-				keyAccountId: *identity.Account,
+				keyAccountId: AccountId,
```
Comment (sergenyalcin) [inline internal/clients/aws.go:92]:
```suggestion
				keyAccountId: accountId,
```

**Review by sergenyalcin — COMMENTED** [review]
Thanks @alexinthesky for this PR. I left a few comments.

**Comment by sergenyalcin** [comment]
Let me also trigger an end-to-end test for being sure that we do not break anything.

**Comment by sergenyalcin** [comment]
/test-examples="examples/ec2/vpc.yaml"

**Comment by sergenyalcin** [comment]
From the logs of end-to-end tests, the provider panics because of the nil pointer error:

`Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference`

```
github.com/upbound/provider-aws/internal/clients.TerraformSetupBuilder.func1({0x3ccbb90, 0xc004092480}, {0x3cd4fe8, 0xc0008b4870}, {0x3d449c0?, 0xc003bd2d80?})
	github.com/upbound/provider-aws/internal/clients/aws.go:64 +0x455
```

The error message points to this line: `github.com/upbound/provider-aws/internal/clients/aws.go:64` 

```go
if len(pc.Spec.Endpoint.Services) > 0 && *pc.Spec.Endpoint.URL.Static == "" {
			return terraform.Setup{}, errors.Wrap(err, "endpoint is wrong")
		}
```

Before the `if` block, we need to check `*pc.Spec.Endpoint.URL.Static` if this pointer is nil or not.

**Comment by turkenf** [comment]
Require status checks changed could you please rebase this PR?

**Comment by alexinthesky** [comment]
/test-examples="examples/ec2/vpc.yaml"

**Comment by alexinthesky** [comment]
@turkenf took the comments and rebased good to go?

**Comment by jeanduplessis** [comment]
@alexinthesky there's a remaining conflict on the PR to be resolved.
If you're interested in pursuing this PR further, please rebase on the latest version in the main branch.

**Comment by alexinthesky** [comment]
Hi, I just rebased hope this makes it upstream !

**Comment by alexinthesky** [comment]
hi here can we merge plz

**Comment by alexinthesky** [comment]
@jeanduplessis

**Comment by jeanduplessis** [comment]
@alexinthesky we are preparing to do new provider releases early next week. We'll include this PR in the release.

File: apis/v1beta1/types.go [hunk 8]
```diff
@@ -21,6 +21,21 @@ type ProviderConfigSpec struct {
 	// of AWS calls made by the provider.
 	// +optional
 	Endpoint *EndpointConfig `json:"endpoint,omitempty"`
+	// Whether to skip credentials validation via the STS API.
+	// This can be useful for testing and for AWS API implementations that do not have STS available.
+	// +optional
+	SkipCredsValidation bool `json:"skip_credentials_validation,omitempty"`
```
Comment (clementblaise) [inline apis/v1beta1/types.go:27]:
```suggestion
	SkipCredsValidation bool `json:"skipCredentialsValidation,omitempty"`
```

**Comment by clementblaise** [comment]
I just noticed that the `ProviderConfig` JSON parameters are defined as snake case, usually in Kubernetes APIs use camel

**Comment by sergenyalcin** [comment]
/test-examples="examples/ec2/vpc.yaml"

**Comment by sergenyalcin** [comment]
From the logs of end-to-end tests, still the provider panics because of the nil pointer error. Also we need to integrate this change to the new code-base.

We want to include this change to the new release. So we will try to handle them.

**Comment by sergenyalcin** [comment]
/test-examples="examples/ec2/vpc.yaml"

**Comment by arielsrv** [comment]
I have the same issue on v.0.35.0, It's works?

**Pull request merged**